cranelift/isle/isle/src/serialize.rs

```diff
@@ -372,10 +390,19 @@ impl<'a> Decomposition<'a> {
         // no unhandled rules left, or because every candidate for control flow
         // for the remaining rules has already been matched by some ancestor in
         // the tree.
-        debug_assert!(self
-            .candidates
-            .iter()
-            .all(|c| c.state == BindingState::Matched));
+        debug_assert!(self.candidates.iter().all(|c| {
```

Comment:
Could we have a bit more on the comment above describing why we are asserting something about equality-check candidates in particular below?